### PR TITLE
Leverage trust domain in issuing certs; remove TD from identity

### PR DESCRIPTION
**Pull request by steeling**
This PR finalizes the steps to allow for custom trust domains. It accomplishes this by:

1) Remove trust domain from service identity, since trust domain can now change, and we are decoupling identities from trust domain, [see this doc for more info](https://docs.google.com/document/d/1nLI4PcfB7-hdPvJipQwQ9dw28f8G7p77iqv3VzsCRFI/edit)
2) Modify IssueCertificate to accept a CN prefix instead of a full CN, and apply the trust domain dynamically. However there are a few cases where a custom trust domain *cannot* be used (k8s webhooks are required to end in `.svc`). So we further modify the IssueCertificate method to accept Variadic args. This neatly fits with the Validation period which is only needed for control plane certs, and defaults to the cached validity duration (cached due to cert rotation).

So we add another variadic arg that signals that the provided prefix is actually the entire common name, and not to append the trust domain.

The remaining changes are simply plumbing around removing .cluster.local from service identities, and calls to IssueCertificate

### CODE REVIEW

File: pkg/envoy/sds/response.go [hunk 1]
```diff
@@ -37,7 +38,7 @@ func NewResponse(meshCatalog catalog.MeshCataloger, proxy *envoy.Proxy, request
 	log.Info().Str("proxy", proxy.String()).Msgf("Creating SDS response for request for resources %v", requestedCerts)
 
 	// 1. Issue a service certificate for this proxy
-	cert, err := certManager.IssueCertificate(certificate.CommonName(s.serviceIdentity), cfg.GetServiceCertValidityPeriod())
+	cert, err := certManager.IssueCertificate(s.serviceIdentity.String())
```
Comment (nojnhuh) [inline pkg/envoy/sds/response.go:41]: Is there a reason we're not setting a validity duration on this cert? Do we already know the cert exists which basically turns this into a cache lookup and the cert will already have one defined?

File: pkg/certificate/manager.go [hunk 2]
```diff
@@ -112,9 +128,15 @@ func (m *Manager) getFromCache(cn CommonName) *Certificate {
 }
 
 // IssueCertificate implements Manager and returns a newly issued certificate from the given client.
-func (m *Manager) IssueCertificate(cn CommonName, validityPeriod time.Duration) (*Certificate, error) {
+func (m *Manager) IssueCertificate(prefix string, opts ...IssueOption) (*Certificate, error) {
 	var err error
-	cert := m.getFromCache(cn) // Don't call this while holding the lock
+	cert := m.getFromCache(prefix) // Don't call this while holding the lock
+
+	options := defaultOptions(m.serviceCertValidityDuration)
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/certificate/manager.go:135]: So there's 2 types here, an `IssueOption`, which is a type alias for a function that operates on a `type issueOptions struct`

The manager has a service validity duration to leverage specifically for rotation, as it doesn't have that info at rotation time otherwise. Here, we create an `issueOptions` type, using the default validity duration, and then apply all of the `IssueOption`, which could override the validity period to something else.

Note the only exported type is IssueOption, so the defaultOptions/issueOptions are a hidden implementation detail.

File: pkg/envoy/sds/response.go (hunk 1 above)
Comment (steeling) [reply pkg/envoy/sds/response.go:41]: The manager has a default validity period it uses, which we're getting from the MRC, so won't have access through the configurator in the future.

Removed configurator from the sdsimpl

File: pkg/ingress/gateway.go [hunk 3]
```diff
@@ -197,8 +197,7 @@ func (c *client) removeGatewayCertAndSecret(storedCertSpec configv1alpha2.Ingres
 		return err
 	}
 
-	certCN := certificate.CommonName(storedCertSpec.SubjectAltNames[0]) // Only single SAN is supported in certs
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/ingress/gateway.go:200]: added back

File: pkg/trafficpolicy/types.go [hunk 4]
```diff
@@ -59,10 +59,11 @@ type InboundTrafficPolicy struct {
 	Rules     []*Rule  `json:"rules:omitempty"`
 }
 
-// Rule is a struct that represents which service identities (authenticated principals) can access a Route
+// Rule is a struct that represents which authenticated principals can access a Route.
+// A principal is of the form <service-identity>.<trust-domain>. It can also contain wildcards.
 type Rule struct {
 	Route                    RouteWeightedClusters `json:"route:omitempty"`
-	AllowedServiceIdentities mapset.Set            `json:"allowed_service_identities:omitempty"`
+	AllowedServiceIdentities mapset.Set            `json:"allowed_princinpals:omitempty"`
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/trafficpolicy/types.go:66]: only if we ever log it maybe? and even then i'm not sure if the logger we use attempts to marshal objects to json.

@shashankram do you know?

File: pkg/certificate/manager.go [hunk 5]
```diff
@@ -60,14 +60,23 @@ func (m *Manager) Start(checkInterval time.Duration, stop <-chan struct{}) {
 // Note that the CRD uses a default, so this value will always be set.
 func (m *Manager) GetTrustDomain() string {
 	// TODO(4754): implement
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/certificate/manager.go]: yup, removed the todo

**Comment by shashankram** [comment]
Could you describe in the commit why the trust domain is being removed from the identity?

**Comment by steeling** [comment]
> Could you describe in the commit why the trust domain is being removed from the identity?

done!

File: pkg/catalog/ingress.go [hunk 6]
```diff
@@ -124,6 +124,7 @@ func (mc *MeshCatalog) GetIngressTrafficPolicy(svc service.MeshService) (*traffi
 				HTTPRouteMatch:   trafficpolicy.WildCardRouteMatch,
 				WeightedClusters: mapset.NewSet(backendCluster),
 			},
+			// TODO(steeling): one place its set
```
Comment (shashankram) [inline pkg/catalog/ingress.go:127]: I don't understand this comment. Could you rephrase or remove this?
Also, it's preferable for TODO comments to reference an issue, e.g. `TODO(#<issue>): ...`

File: pkg/trafficpolicy/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/trafficpolicy/types.go:75]: We probably don't need to have JSON tags for internal structs in most cases, except when the type is used with the debugger pkg to render JSON. If these types are not used by the debugger pkg, we can remove them. In most cases, it exists because of following a pattern and for no apparent reason.

**Review by shashankram — APPROVED** [review]

File: pkg/catalog/ingress.go (hunk 6 above)
Comment (steeling) [reply pkg/catalog/ingress.go:127]: that was a personal note that snuck in, removed!

**Comment by jaellio** [comment]
@steeling I can merge this once the failing lint and tests are fixed. Thanks for this change!
